### DISCUSSION TITLE
Apply node selector to ArgoCD

### DISCUSSION THREAD

**Discussion by AlessioCasolaro** [argoproj/argo-cd#18984] [Q&A]
Hi, I am writing since I would like to assign ArgoCD deployments on a K8s node selector. I was following the guide [ArgoCD Nodeplacement](https://argocd-operator.readthedocs.io/en/latest/reference/argocd/#nodeplacement-option) so I tried to format the message by entering:
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: nodeplacement-argocd
  namespace: my-namespace
spec:
  nodePlacement:
    nodeSelector:
      name: my-node
```

When I try `kubectl apply -n my-namespace -f nodeplacement.yaml` I got
```
error: resource mapping not found for name: "nodeplacement-argocd" namespace: "my-namespace" from "nodeplacement.yaml": no match for type "ArgoCD" in version "argoproj.io/v1alpha1".
make sure CRDs are installed first
```

If I run ` kubectl get crds | grep argoproj.io` I got

```
applications.argoproj.io 2023-11-29T17:04:37Z
applicationsets.argoproj.io 2023-11-29T17:04:37Z
appprojects.argoproj.io 2023-11-29T17:04:36Z
argocdextensions.argoproj.io 2023-12-21T10:24:40Z
```

How I can solve the problem?
Thanks in advance!

**Comment by ChristianCiach** [comment]
You should probably ask this over at the project of the Argo-CD operator: https://github.com/argoproj-labs/argocd-operator

**Reply by ChristianCiach** [reply]
Which, funnily enough, is probably the answer to your question. You probably deployed Argo-CD (from this repo) directly, instead of deploying the operator?

**Reply by AlessioCasolaro** [reply]
> Which, funnily enough, is probably the answer to your question. You probably deployed Argo-CD (from this repo) directly, instead of deploying the operator?

Yes I deployed argocd directly from Bitnami, at the moment I don't have the controller. Is it possible to install the controller as a stand-alone entity that connects with the existing argocd deployment?

**Reply by ChristianCiach** [reply]
> Yes I deployed argocd directly from Bitnami

Sorry, I don't even know what this means. I guess you used the "Bitnami package" of argo-cd, which is essentially just a helm chart?

If so, then you can just add the node selectors/affinities to your `values.yaml`: https://github.com/bitnami/charts/blob/a26e7792554e763473efdb9da309c4eaa4f36edb/bitnami/argo-cd/values.yaml#L556

I am pretty sure that you actually don't want to deploy the Argo-CD Operator. It is experimental anyway.

**Reply by AlessioCasolaro** [reply]
> > Yes I deployed argocd directly from Bitnami
> 
> Sorry, I don't even know what this means. I guess you used the "Bitnami package" of argo-cd, which is essentially just a helm chart?
> 
> If so, then you can just add the node selectors/affinities to your `values.yaml`: https://github.com/bitnami/charts/blob/a26e7792554e763473efdb9da309c4eaa4f36edb/bitnami/argo-cd/values.yaml#L556
> 
> I am pretty sure that you actually don't want to deploy the Argo-CD Operator. It is experimental anyway.

Yes I have used the Bitnami package of argo-cd

So it should be a file `values.yaml` with 
```
nodeSelector:
  name: my-node
```

an then `kubectl apply -n my-namespace -f values.yaml` right? Thanks in advance for your help!!

**Reply by AlessioCasolaro** [reply]
> > Yes I deployed argocd directly from Bitnami
> 
> Sorry, I don't even know what this means. I guess you used the "Bitnami package" of argo-cd, which is essentially just a helm chart?
> 
> If so, then you can just add the node selectors/affinities to your `values.yaml`: https://github.com/bitnami/charts/blob/a26e7792554e763473efdb9da309c4eaa4f36edb/bitnami/argo-cd/values.yaml#L556
> 
> I am pretty sure that you actually don't want to deploy the Argo-CD Operator. It is experimental anyway.

I'm probably still missing a few pieces to reconnect, maybe my value.yaml example is wrong, could you help me figure out how I can make one to apply the node selector? It is not very clear to me.

**Reply by ChristianCiach** [reply]
Please describe to me how you installed Argo-CD in the first place. Maybe this will make things clearer to me.

**Reply by AlessioCasolaro** [reply]
> Please describe to me how you installed Argo-CD in the first place. Maybe this will make things clearer to me.

Sure, I try to provide you with any details you may need. I installed argocd with the Bitnami package found at the link https://bitnami.com/stack/argo-cd/helm

**Reply by ChristianCiach** [reply]
So you surely used a command like `helm install ...` to install the Helm chart, right? In this case I would refer you to the official Helm documentation about how to provide your custom configuration values to `helm install` or `helm upgrade`:

* https://helm.sh/docs/chart_template_guide/values_files/

These are the possible configuration values for the Helm chart from Bitnami:

* https://github.com/bitnami/charts/blob/main/bitnami/argo-cd/values.yaml

(Off-Topic, but I would probably prefer the [official helm chart](https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd) over the Bitnami one.)

As an example, you could create a values-file like `my-argocd-values.yaml` with this content:

```yaml
controller:
  nodeSelector:
    some-node-label: some-value

applicationSet:
  nodeSelector:
    some-node-label: some-value

notifications:
  nodeSelector:
    some-node-label: some-value

server:
  nodeSelector:
    some-node-label: some-value

repoServer:
  nodeSelector:
    some-node-label: some-value

dex:
  nodeSelector:
    some-node-label: some-value
```

You could then use Helm to install/upgrade Argo-CD using this file:

```
helm install -f my-argocd-values.yaml ...
```

But nothing about this has anything to do with Argo-CD. You should read up about how to use Helm.

**Reply by AlessioCasolaro** [reply]
> So you surely used a command like `helm install ...` to install the Helm chart, right? In this case I would refer you to the official Helm documentation about how to provide your custom configuration values to `helm install` or `helm upgrade`:
> 
> * https://helm.sh/docs/chart_template_guide/values_files/
> 
> These are the possible configuration values for the Helm chart from Bitnami:
> 
> * https://github.com/bitnami/charts/blob/main/bitnami/argo-cd/values.yaml
> 
> (Off-Topic, but I would probably prefer the [official helm chart](https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd) over the Bitnami one.)
> 
> As an example, you could create a values-file like `my-argocd-values.yaml` with this content:
> 
> ```yaml
> controller:
>   nodeSelector:
>     some-node-label: some-value
> 
> applicationSet:
>   nodeSelector:
>     some-node-label: some-value
> 
> notifications:
>   nodeSelector:
>     some-node-label: some-value
> 
> server:
>   nodeSelector:
>     some-node-label: some-value
> 
> repoServer:
>   nodeSelector:
>     some-node-label: some-value
> 
> dex:
>   nodeSelector:
>     some-node-label: some-value
> ```
> 
> You could then use Helm to install/upgrade Argo-CD using this file:
> 
> ```
> helm install -f my-argocd-values.yaml ...
> ```
> 
> But nothing about this has anything to do with Argo-CD. You should read up about how to use Helm.

Okay thank, you helped me so much to understand.
Other question do you happen to know if with the same approach I can switch from ArgoCD in its "classic" version to ArgoCD High Availability?
I guess between the two versions there are only differences in terms of parameters to configure, so I could follow this helm value https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/README.md#high-availability and apply it to my current argo deployment?

**Reply by ChristianCiach** [reply]
I didn't try it, but upgrading from non-HA to HA deployment *should* be possible. Argo-CD is stateless, so the only difference should be the Redis cache (which can be thrown anyway at any time without losing anything important).

But you linked to the *official* Argo-CD helm chart. Do not confuse the two helm charts; they are different. If you want to migrate to the official helm chart, you should uninstall the Bitnami one first.

**Reply by AlessioCasolaro** [reply]
> I didn't try it, but upgrading from non-HA to HA deployment _should_ be possible. Argo-CD is stateless, so the only difference should be the Redis cache (which can be thrown anyway at any time without losing anything important).
> 
> But you linked to the _official_ Argo-CD helm chart. Do not confuse the two helm charts; they are different. If you want to migrate to the official helm chart, you should uninstall the Bitnami one first.

Yes I understood that in the actual they are two different packages the official one versus the bitnami one, I was seeing though that most of the configurations given at the link I mentioned (https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/README.md#high-availability) , seem to be there also in the values file of Bitnami (https://github.com/bitnami/charts/blob/main/bitnami/argo-cd/values.yaml) and so they should be settable. The only one I am missing  is 
```
redis-ha:
  enabled: true 
```

Do you happen to know how I could move?
